lib/blockchain.go

```diff
@@ -558,26 +628,27 @@ func (bc *Blockchain) _initChain() error {
 	// add a block's parents, if they exist, before adding the block itself.
 	var err error
 	if bc.postgres != nil {
-		bc.blockIndex, err = bc.postgres.GetBlockIndex()
+		bc.blockIndexByHash, err = bc.postgres.GetBlockIndex()
 	} else {
-		bc.blockIndex, err = GetBlockIndex(bc.db, false /*bitcoinNodes*/)
+		bc.blockIndexByHash, err = GetBlockIndex(bc.db, false /*bitcoinNodes*/)
 	}
 	if err != nil {
 		return errors.Wrapf(err, "_initChain: Problem reading block index from db")
 	}
+	bc.blockIndexByHeight = bc.ConstructBlockIndexByHeight()
 
-	// At this point the blockIndex should contain a full node tree with all
+	// At this point the blockIndexByHash should contain a full node tree with all
 	// nodes pointing to valid parent nodes.
 	{
 		// Find the tip node with the best node hash.
-		tipNode := bc.blockIndex[*bestBlockHash]
+		tipNode := bc.blockIndexByHash[*bestBlockHash]
 		if tipNode == nil {
 			return fmt.Errorf("_initChain(block): Best hash (%#v) not found in block index", bestBlockHash)
 		}
 
 		// Walk back from the best node to the genesis block and store them all
 		// in bestChain.
-		bc.bestChain, err = GetBestChain(tipNode, bc.blockIndex)
+		bc.bestChain, err = GetBestChain(tipNode, bc.blockIndexByHash)
```

Comment:
So the uncommitted blocks that are after the committed tip and validated aren't added to best chain? How do they get added?

Seems like maybe you could add them here now that you have blockIndexByHeight.

Reply:
Once we have a new block come in, the process block PoS logic automatically determines where to reorg to. We can add some extra logic here to update the best chain on restart, but this is a minor improvement as we'll auto re-org when we get the first new block.